lightning/src/routing/gossip.rs

```diff
@@ -2532,32 +2533,55 @@ mod tests {
+		gossip_sync.handle_channel_update(&valid_channel_update_2).unwrap();
+		assert!(network_graph.read_only().channels().get(&short_channel_id).unwrap().two_to_one.is_some());
+
 		network_graph.remove_stale_channels_and_tracking_with_time(100 + STALE_CHANNEL_UPDATE_AGE_LIMIT_SECS);
 		assert_eq!(network_graph.read_only().channels().len(), 1);
 		assert_eq!(network_graph.read_only().nodes().len(), 2);
 
 		network_graph.remove_stale_channels_and_tracking_with_time(101 + STALE_CHANNEL_UPDATE_AGE_LIMIT_SECS);
+		#[cfg(not(feature = "std"))] {
+			// Make sure removed channels are tracked.
+			assert_eq!(network_graph.removed_channels.lock().unwrap().len(), 1);
+		}
+
 		#[cfg(feature = "std")]
 		{
 			// In std mode, a further check is performed before fully removing the channel -
 			// the channel_announcement must have been received at least two weeks ago. We
-			// fudge that here by indicating the time has jumped two weeks. Note that the
-			// directional channel information will have been removed already..
+			// fudge that here by indicating the time has jumped two weeks.
 			assert_eq!(network_graph.read_only().channels().len(), 1);
 			assert_eq!(network_graph.read_only().nodes().len(), 2);
-			assert!(network_graph.read_only().channels().get(&short_channel_id).unwrap().one_to_two.is_none());
 
+			// Note that the directional channel information will have been removed already..
+			// We want to check that this will work even if *one* of the channel updates is recent,
+			// so we should add it with a recent timestamp.
+			assert!(network_graph.read_only().channels().get(&short_channel_id).unwrap().one_to_two.is_none());
 			use std::time::{SystemTime, UNIX_EPOCH};
 			let announcement_time = SystemTime::now().duration_since(UNIX_EPOCH).expect("Time must be > 1970").as_secs();
+			let valid_channel_update = get_signed_channel_update(|unsigned_channel_update| {
+				unsigned_channel_update.timestamp = (announcement_time + 1 + STALE_CHANNEL_UPDATE_AGE_LIMIT_SECS) as u32;
+			}, node_1_privkey, &secp_ctx);
+			assert!(gossip_sync.handle_channel_update(&valid_channel_update).is_ok());
+			assert!(network_graph.read_only().channels().get(&short_channel_id).unwrap().one_to_two.is_some());
 			network_graph.remove_stale_channels_and_tracking_with_time(announcement_time + 1 + STALE_CHANNEL_UPDATE_AGE_LIMIT_SECS);
+			// Make sure removed channels are tracked.
+			assert_eq!(network_graph.removed_channels.lock().unwrap().len(), 1);
+			// Provide a later time so that sufficient time has passed
+			network_graph.remove_stale_channels_and_tracking_with_time(announcement_time + 1 + STALE_CHANNEL_UPDATE_AGE_LIMIT_SECS +
+				REMOVED_ENTRIES_TRACKING_AGE_LIMIT_SECS);
 		}
 
 		assert_eq!(network_graph.read_only().channels().len(), 0);
 		assert_eq!(network_graph.read_only().nodes().len(), 0);
+		assert!(network_graph.removed_channels.lock().unwrap().is_empty());
```

Comment:
Hmm. Do we even need the std block above. That comment saying the extra check only for std doesn't seem accurate really as the remove_stale* method doesn't have any std vs no-std discrimination in its body. Maybe it's old? It seems that if we just don't use SystemTime in the block above and just use some hardcoded announcement_time in seconds and everything works fine.

Reply:
By looking the code, you can see the difference by looking that the
`assert_eq!(network_graph.read_only().channels().len(), 0);` check.

For no-std, this is true immediately (after std-block).
For std, it's false initially (see the first check inside the block), and then becomes true after some more manipulations.

--------------

If you prefer a white-box description, the difference I think is `info.announcement_received_time`, which has a real timestamp for std, and which we don't have control of. So that's why an extra block is needed.